interfaces/repo.go

```diff
@@ -516,3 +516,110 @@ func (r *Repository) securitySnippetsForSnap(snapName string, securitySystem Sec
 	}
 	return snippets, nil
 }
+
+// AddSnap adds plugs and slots declared by the given snap to the repository.
+//
+// This function can be used to implement snap install or, when used along with
+// RemoveSnap, snap upgrade.
+//
+// Plugs/slots that are already present in the repository are not changed. Only
+// new plugs/slots are added.
+//
+// Each added plug/slot is validated according to the corresponding interface.
+// Unknown interfaces and plugs/slots that don't validate are not added.
+// Information about those failures are returned to the caller.
+func (r *Repository) AddSnap(snapInfo *snap.Info) []error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	snapName := snapInfo.Name
+
+	for plugName := range snapInfo.Plugs {
+		if _, ok := r.plugs[snapName][plugName]; ok {
+			panic(fmt.Errorf("cannot add plug %s.%s: existing plug in the way", snapName, plugName))
```

Comment:
Yeah, I think you are right. There are times where panic() is useful but here it's really bringing down the whole system. While invalid use of this function is still a bug it's better to handle it at least somewhat gracefully.

Both functions no longer panic.